java/ql/src/semmle/code/java/dataflow/NullGuards.qll

```diff
@@ -44,21 +44,32 @@ Expr clearlyNotNullExpr(Expr reason) {
   or
   result instanceof ArrayCreationExpr and reason = result
   or
+  result instanceof FunctionalExpr and reason = result
+  or
   result instanceof TypeLiteral and reason = result
   or
   result instanceof ThisAccess and reason = result
   or
   result instanceof StringLiteral and reason = result
   or
+  // Add and AssignAdd performing String concatenation never have null result
   result instanceof AddExpr and result.getType() instanceof TypeString and reason = result
   or
+  result instanceof AssignAddExpr and result.getType() instanceof TypeString and reason = result
+  or
   exists(Field f |
     result = f.getAnAccess() and
-    (f.hasName("TRUE") or f.hasName("FALSE")) and
+    f.hasName(["TRUE", "FALSE"]) and
     f.getDeclaringType().hasQualifiedName("java.lang", "Boolean") and
     reason = result
   )
   or
+  result.getType() instanceof PrimitiveType and not result instanceof TypeAccess and reason = result
```

Comment:
These cannot be null simply due to their type, so possibly out of scope for this predicate. But I could imagine places where this disjunct might contribute. Please explain where you think that's the case and consider whether those cases are handled better without including all primitive-type expressions in this predicate.

Reply:
> These cannot be null simply due to their type

Can't they? At least for expressions which actually have a value a primitive value can never be `null` (any dereferencing would have thrown a `NullPointerException` before).
For non-value expressions this of course does not apply, which is why I excluded `TypeAccess`; maybe there are more which need to be excluded. Though I assume since this predicate is mostly used with control flow analysis, such `Expr` types would not be part of the result set in the first place, but maybe it would be good to restrict it more nonetheless?

Reply:
I agree that this addition is technically correct, hence "These cannot be null simply due to their type". What I'm questioning is the value that this adds, since the cost is a great many tuples.

Reply:
Sorry, misread your comment. Yes, performance considerations might be a fair point (you can probably estimate or meassure that better than me).

However, this change does increase the number of findings, see [this query](https://lgtm.com/query/6737122779327441283/).